### ISSUE TITLE
Front page statistics

### ISSUE THREAD

**Issue by gwicke** [gwicke/testreduce#17]
Maintain the summary statistics currently used to display the front page
- In-memory: keep running averages, ratios of errors / fails / skips etc
- In storage: store summary per revision
- provide a backend interface for this
  - latest statistics (front page)
  - statistics for past revisions (long-term trends)
- hook up the front page to use these back-end interfaces

Who: John, Brando

**Comment by choucalate** [comment]
I have 3 questions:
1) Are the averages, ratios of errors/ fails/ skips all part of one revision?
2) For storing the summary per revision, do we need a new table for this? Or should I bring everything into memory and calculate it?
3) Where is the route for the statistics for past revisions/long-term trends? I don't think it's the green/yellow/red bar right?

**Comment by brando90** [comment]
I am unsure on the exact meaning of some of the statistics we are trying to compute. Is there a formula or a algorithm or something that describes them precisely? 

Some of them are:
-Syntactic diffs  
-Semantic diffs 
-Errors

Also its a little unclear to me of what is _everything_ that we want to compute.

**Comment by gwicke** [comment]
@choucalate: Averages are currently not persisted by revision, but it would be great if we could do so. The front page statistics display dynamically calculated stats based on the latest available information. We currently can't display only long-term trends in averages as dynamic computation of those would be way too expensive. So computing and persisting this would be great. For the latest revision / state it makes sense to keep this updated in memory & save it back to a cassandra table. For old revisions the stats can then be simply loaded from the db table.

@brando90: All stats are defined in the MySQL code, see server-mysql.js. Syntactic / semantic diffs are fails/skips in the JUnit result XML. Errors are errors in the JUnit XML or a test that returned no result within the retry count limit.

**Comment by choucalate** [comment]
I made a pull request https://github.com/gwicke/testreduce/pull/28

**Comment by choucalate** [comment]
I've more or less finished this issue, and I'll be issuing a pull request when I finish some topfails stuff
https://github.com/choucalate/testreduce - if you want to take a look at it

Still need to do before this issue is closed:
Possibly revision summaries (will probably restructure a couple things), 
Crashers
Testing on scalability
Averages

**Comment by brando90** [comment]
@choucalate , are there still things to do here? Do you need any help in anything? I see you have the following list:

Possibly revision summaries (will probably restructure a couple things), 
Crashers
Testing on scalability
Averages

Do you want to split them up or something? That way we don't end up doing the same thing twice or something?

**Comment by choucalate** [comment]
Sure! did you have an idea for how to split it?

**Comment by brando90** [comment]
Hey John!  I think I don't have a good sense what the tasks actually entail, lol, so if I choose how to divide it, it might not be a fair division :p thus, if u feel u might know a better/more fair division of labour, go ahead an assign it. The one task I would not rather do is "Testing on scalability". Any of the other 3 I think sound found. What do you think?

**Comment by jjchen** [comment]
Hey guys! If there's still work to be done here, I can help too.

**Comment by choucalate** [comment]
Hooray, thank you guys for helping!

@brando90 could you implement the current one fail, one skip, other new fails features that exists on the new parsoid site?  you'll have to make some new methods in cassandrabackend to fetch stuff (http://parsoid.wmflabs.org:8001/ ).
and
@jjchen could you do revision summaries and averages?  I also don't have the front end completely set up to do averages yet either.  You can probably just start with running averages and then do revision summaries later.  

if you guys want to switch around roles or do something else that's fine as well

**Comment by jjchen** [comment]
Okay, sounds good. Just to check my understanding: averages are over the scores we have for the latest commit? And are revisions the same as commits?

**Comment by brando90** [comment]
Sure, but how do I know what features we want? Is there documentation for it or should I read some code (which code) ... or is it on the html we create for round trip testing? If I can see somewhere what feature I should implement, I should be able to do it :)

**Comment by jjchen** [comment]
@choucalate It looks like the statistics are computed each time in getStatistics, is that correct? So it seems pretty simple/inexpensive to add an averages calculation in there (since we're looping through all the scores anyways). The alternative might be to keeping a running average and updating in addResult - thoughts? Do we want to keep the results in memory somewhere? Also, I'm thinking of doing the insert into the revision summaries table in the when we see a new commit (in updateCommits).

**Comment by jjchen** [comment]
@brando90 I think @choucalate is talking about the "Flagged Regressions" links here: http://parsoid.wmflabs.org:8001/ which look like it's comparing the current commit results with results from the previous commit.

**Comment by jjchen** [comment]
Also, just poking around the links on the front page, it looks like the "Non-existing test pages", "Histogram of failures", and "Histogram of skips" pages still need to be implemented. I can start that if nobody's working on it yet (maybe it should be a separate issue?).

**Comment by jjchen** [comment]
So I have a 'working' version of the averages/summary (https://github.com/jjchen/testreduce/commit/03c6860cee15191b24f3c347b1364d22e3d39c40) and histogram of failures/skips (https://github.com/jjchen/testreduce/commit/37327aa5a411f541308d6d9889e93ac86615bc52), but I think I'm going to change the averages to be updated in addResult (unless there's a compelling reason not to?) and I need to do some refactoring. 

One problem I'm having though is the score to skips/errors/fails conversion. We use errorCount_1000000+failCount_1000+skipCount to get the score, & the inverse to get the skips/errors/fails, but if the skipCount was originally > 999, wouldn't that get counted as a fail?

**Comment by choucalate** [comment]
@jjchen yah i think that's right, commit summaries = revision summaries
As for your second post, YUP! i think doing an averages calculation in getStats would probably be the best idea for now and doing an insert into the revision summaries table.  Just making sure, your insert is when you see a new commit and you insert the old commit summaries?  

Yah I haven't even looked at the non-existing test pages and stuffs yet, perhaps a new issue would be best.

@brando90 yup that's what I meant, sorry I forgot the link!

**Comment by choucalate** [comment]
@jjchen for the stats thing... yah pretty much lol. There's no way to invert the scores as of right now, unless we change the scoring method.  However, none of the "current" ones go over 999... so safe for now? lol ...
right now the worst is 823 skips- http://parsoid.wmflabs.org:8001/topfails/13

the only solution i can think of is either readjusting the scoring method or saving error fail and skips inside test_By_score

**Comment by jjchen** [comment]
@choucalate yeah, that's what I thought (for the scores thing) but then I was looking at the skips page (http://parsoid.wmflabs.org:8001/skipsDistr) and it seems to go to 1734? idk if that's a big problem?

**Comment by choucalate** [comment]
ohh woops... hahah totally missed that. mmm any other ideas to extract data from score?

**Comment by jjchen** [comment]
haha yeah, so as you said, we could just readjust the scoring method -- which might not be a real fix, but something like failCount \* 10,000 should be pretty safe?

**Comment by brando90** [comment]
@choucalate I want to make sure I understand what is suppose to be done. Is it similar to your getStatistics function?

For the current commit, do I loop through the test_by_score table and for rows that match the current commit I figure out if it has a single fail or a single skip? Once I figure out which ones have a single skip or fail, what do I do with them? Feed them to server.js supplied callback? Or insert them to the the database? Both?

**Comment by choucalate** [comment]
Sorry for the late response, I've been pretty busy finishing up some projects
@jjchen yah i think that would probably be the best quick fix
@brando90 Hm actually it might be easier than that:
Correct me if i'm wrong, but i think each of them are just subsets of the regressions between the current commit and the one right before

I think you just have to call the cassandrabackend method for retrieving the regressions and do some filtering?

**Comment by brando90** [comment]
@choucalate don't worry, I understand, I've also been busy with projects, hehe. 

Anyway, to get the information for a single fail or a single skip, it seems to me that the only current table that seems to hold that type of info is your new revision_summary table. Thus, to get that info kind of information about a regression of two commits, is the information about the specific commits inside the revision blob? Or am I completely off?

Also, to figure out what "current" commit means and "old" means, is current just the one received as an argument and the previous one, we figure it out by time stamp?

Thanks so much for discussing this with me :)

**Comment by choucalate** [comment]
@brando90 I think the revision table just contains a summary (as far as i know).  You have to get the raw data from it.  Luckily, this function- calcregressionfixes, within 

https://github.com/choucalate/testreduce/blob/master/CassandraBackend.js#L615

does that  for you.  You can just use the method and then extract all the ones where there only one skip, one fail, and etc.  

But in a general sense your method should just be one that accepts 2 inputs, which are the commits, and returns information in a callback about those 3 parts.  
the route will be something like this
http://parsoid.wmflabs.org:8001/(oneofthethree)/between/commit1/commit2
however, neither commit has to be earlier or later than another, you should simply calculate the data required for them.

**Comment by brando90** [comment]
@choucalate @jjchen Thanks so much guys! I think I know what to do now. Let me give it a try now :)

**Comment by brando90** [comment]
I believe I am approaching finishing my task, is there anything else you guys had in mind that I could work on? @jjchen @choucalate

**Comment by brando90** [comment]
@jjchen @choucalate  I currently don't really have a task to work on. If you guys want me to work on anything specific, feel free to tell me. :)

**Comment by choucalate** [comment]
yah im also waiting on a task, im not exactly sure which part to work on either

**Comment by choucalate** [comment]
Made some quick fixes...  not sure what to do next.
Is there something to do?
